# Expenses in Highlands Sheet are incorrect
* Using expenses from Data link to fill in expenditures for March 25 presentation


In [1]:
import numpy as np
import pandas as pd
import TIRCP_functions
from calitp_data_analysis.sql to_snakecase

GCS_FILE_PATH = "gs://calitp-analytics-data/data-analyses/tircp/"

pd.options.display.max_columns = 50
pd.options.display.max_rows = 120
pd.set_option('display.max_colwidth', None)

/opt/conda/lib/python3.9/site-packages/geopandas/_compat.py:111: UserWarning: The Shapely GEOS version (3.10.2-CAPI-1.16.0) is incompatible with the GEOS version PyGEOS was compiled with (3.10.1-CAPI-1.16.0). Conversions between both will be slow.
E0316 20:20:50.324598556    1219 fork_posix.cc:70]           Fork support is only compatible with the epoll1 and poll polling strategies
E0316 20:20:50.840796541    1219 fork_posix.cc:70]           Fork support is only compatible with the epoll1 and poll polling strategies


## Expenditure information from the Enterprise Data Link data Linda pulled.

In [2]:
enterprise = "Enterprise.xlsx"
enterprise_df = pd.read_excel(f"{GCS_FILE_PATH}{enterprise}")
enterprise_df = to_snakecase(enterprise_df)

In [3]:
enterprise_df.head(2)

fy cycle  fund  appr_catg appr_unit      pec   project  \
0  2016   One  3228       1415     15101  3010070  16000007   
1  2016   One  3228       1415     15101  3010070  16000008   

                                 project_name project_short_name phase  \
0  Refurbishment of Seven Light Rail Vehicles          64-R340GA     S   
1   South Bay Bus Rapid Transit (BRT) Project          64-T339GA     S   

   old_ea  capital_exp   local_exp  support_exp  other_exp     tot_exp  
0  R340GA            0  2182302.44            0          0  2182302.44  
1  T339GA            0    30770.46            0          0    30770.46

In [4]:
enterprise_df = enterprise_df.rename(columns = {'project':'project_id'})

In [5]:
#keep only columns necessary.
enterprise_df = enterprise_df[['fy','cycle','project_id','project_name','tot_exp']]

In [6]:
#Looking at IDS by cycles.
enterprise_df.cycle.value_counts()

Three    152
One       47
Two       47
Four       8
Name: cycle, dtype: int64

In [7]:
enterprise_df.shape

(254, 5)

#### Aggregate by project ID and each cycle, so each project has only one record for total expenses

In [8]:
enterprise_df_project_id = enterprise_df.groupby(['project_id','cycle']).agg({'tot_exp':'sum'}).reset_index()

In [9]:
enterprise_df_project_id.shape

(106, 3)

In [10]:
enterprise_df_project_id.head(2)

project_id cycle     tot_exp
0    16000007   One  6315208.89
1    16000008   One  4000000.00

## Merge in with allocation sheet of Excel workbook 
* There are 191 unique project IDS in our allocation sheet compared to 106 project IDS in Data Link.

In [11]:
#read in allocation df because that's the sheet with project ids. 
allocation = TIRCP_functions.allocation()

In [12]:
#keep only columns that are relevant.
allocation_df = allocation[['Award_Year','Project_ID','PPNO']]

In [13]:
#checking NA values
allocation_df.isna().sum()

Award_Year    2
Project_ID    7
PPNO          2
dtype: int64

In [14]:
f'There are {allocation_df.Project_ID.nunique()} unique project ids in allocation'

'There are 191 unique project ids in allocation'

In [15]:
#only keep rows that have Project IDS filled in
allocation_df = allocation_df.loc[allocation_df['Project_ID'].notnull()]

In [16]:
allocation_df.shape

(208, 3)

In [17]:
joined_alloc = pd.merge(allocation_df, enterprise_df_project_id, 
                        left_on='Project_ID', right_on='project_id', 
                        how = 'outer', indicator = True)

In [18]:
joined_alloc._merge.value_counts()

both          119
left_only      89
right_only      1
Name: _merge, dtype: int64

In [19]:
#only keep "both" results...
joined_alloc2 = joined_alloc.loc[joined_alloc['_merge'] == 'both'].drop(columns =['_merge'])

In [20]:
#group by so that only one row for each PPNO number & Year
joined_alloc3 = (joined_alloc2
                  .groupby(['PPNO', 'Award_Year'])
                  .agg({'tot_exp':'sum'})
                  .reset_index()
                 )

In [21]:
joined_alloc3.shape

(51, 3)

In [22]:
#rename cols before joining with "joined" df 
joined_alloc3 = joined_alloc3.add_prefix('from_joined_alloc_df_')

In [23]:
joined_alloc3

from_joined_alloc_df_PPNO  from_joined_alloc_df_Award_Year  \
0                       1155                           2018.0   
1                       1230                           2016.0   
2                      2320B                           2018.0   
3                      CP001                           2015.0   
4                      CP002                           2015.0   
5                      CP003                           2015.0   
6                      CP004                           2015.0   
7                      CP005                           2015.0   
8                      CP006                           2015.0   
9                      CP006                           2016.0   
10                     CP006                           2018.0   
11                     CP007                           2015.0   
12                     CP008                           2015.0   
13                     CP010                           2015.0   
14                     CP011                           2015.0   
15                     CP012                           2015.0   
16                     CP013                           2015.0   
17                     CP014                           2015.0   
18                     CP015                           2015.0   
19                     CP016                           2016.0   
20                     CP017                           2016.0   
21                     CP018                           2016.0   
22                     CP019                           2016.0   
23                     CP020                           2016.0   
24                     CP021                           2016.0   
25                     CP022                           2016.0   
26                     CP023                           2016.0   
27                     CP024                           2016.0   
28                     CP025                           2016.0   
29                     CP027                           2018.0   
30                     CP028                           2018.0   
31                     CP031                           2018.0   
32                     CP032                           2018.0   
33                     CP033                           2018.0   
34                     CP034                           2018.0   
35                     CP035                           2018.0   
36                     CP036                           2018.0   
37                     CP039                           2018.0   
38                     CP041                           2018.0   
39                     CP043                           2016.0   
40                     CP046                           2018.0   
41                     CP047                           2018.0   
42                     CP048                           2018.0   
43                     CP051                           2018.0   
44                     CP053                           2018.0   
45                     CP055                           2018.0   
46                     CP061                           2020.0   
47                     CP062                           2020.0   
48                     CP069                           2020.0   
49                     CP072                           2020.0   
50                     CP079                           2018.0   

    from_joined_alloc_df_tot_exp  
0                      486509.70  
1                     9204000.00  
2                      500000.00  
3                     6315208.89  
4                    38023039.68  
5                     4000000.00  
6                     2320000.00  
7                    22284205.53  
8                    41181000.00  
9                    45092000.00  
10                   26867000.00  
11                     277840.47  
12                   31936000.00  
13                   11000000.00  
14                    6841000.00  
15                    4619999.90  
16                   100000

In [24]:
joined_alloc3.astype({'from_joined_alloc_df_Award_Year': 'int64'}).dtypes

from_joined_alloc_df_PPNO           object
from_joined_alloc_df_Award_Year      int64
from_joined_alloc_df_tot_exp       float64
dtype: object

## Joining project sheet on Award Year & PPNO

In [25]:
project = TIRCP_functions.project()

In [26]:
final_join = pd.merge(project, joined_alloc3, 
                      left_on = ['PPNO', 'Award_Year'], 
                      right_on = ['from_joined_alloc_df_PPNO', 'from_joined_alloc_df_Award_Year'],
                      how = 'left', indicator = True)

In [27]:
final_join._merge.value_counts()

both          50
left_only     25
right_only     0
Name: _merge, dtype: int64

#### Check to see if PPNO matches  cross original project sheet & joined 

In [28]:
final_join.shape

(75, 36)

In [29]:
final_join.PPNO.nunique()

69

In [30]:
project.PPNO.nunique()

69

In [31]:
PPNO_project = set(project.PPNO.unique().tolist())
PPNO_join = set(final_join.PPNO.unique().tolist())
PPNO_join - PPNO_project 

set()

In [32]:
pd.options.display.float_format = '{:,}'.format
pd.options.display.float_format = "{:.2f}".format

### Final Clean Up - per Linda, TIRCP isn't correct

In [33]:
#fill in with 0
final_join['from_joined_alloc_df_tot_exp'] = final_join['from_joined_alloc_df_tot_exp'].fillna(0)

In [34]:
final_join = final_join.drop_duplicates(subset=['PPNO','Award_Year', 'Project_Title'])

In [35]:
final_join.loc[(final_join['PPNO'] == 'CP028'), "TIRCP_project_sheet"] =  13156000 
final_join.loc[(final_join['PPNO'] == 'CP019'), "TIRCP_project_sheet"] =  8930000

#### Dropping a project that may or may not be considered TIRCP.

In [36]:
index = final_join[(final_join.PPNO == 'CP026') &( final_join.Award_Year == 2016)].index

In [37]:
final_join = final_join.drop(index)

In [38]:
final_join = final_join.reset_index()

### Checking out our 2 columns of expended amounts to make sure Data Link information makes sense.
* Expended_Amt_Project_sheet is from Highlands sheet.
* from_joined_alloc_df_tot_exp from Data Link

In [39]:
final_join[['Award_Year','PPNO','Project_Title', 'TIRCP_project_sheet', 
            'Expended_Amt_project_sheet', 'from_joined_alloc_df_tot_exp', '_merge']]

Award_Year   PPNO  \
0         2015  CP005   
1         2015  CP012   
2         2015  CP015   
3         2015  CP007   
4         2015  CP013   
5         2015  CP004   
6         2015  CP001   
7         2015  CP003   
8         2015  CP008   
9         2015  CP006   
10        2015  CP014   
11        2015  CP011   
12        2015  CP002   
13        2015  CP010   
14        2016  CP019   
15        2016  CP023   
16        2016  CP076   
17        2016  CP016   
18        2016  CP058   
19        2016  CP020   
20        2016  CP043   
21        2016  CP017   
22        2016  CP022   
23        2016  CP080   
24        2016   1230   
25        2016  CP006   
26        2016  CP025   
27        2016  CP057   
28        2018  2320B   
29        2018  CP027   
30        2018  CP028   
31        2018  CP055   
32        2018  CP036   
33        2018  CP079   
34        2018  CP029   
35        2018  CP054   
36        2018  CP051   
37        2018  CP030   
38        2018  CP031   
39        2018  CP031   
40        2018  CP039   
41        2018  CP053   
42        2018  CP034   
43        2018  CP077   
44        2018  CP034   
45        2018  CP006   
46        2018  CP035   
47        2018  CP078   
48        2018  CP048   
49        2018  CP047   
50        2018  CP057   
51        2018  CP045   
52        2018  CP046   
53        2018  CP041   
54        2018  CP033   
55        2018   1155   
56        2020  CP059   
57        2020  CP060   
58        2020  CP061   
59        2020  CP062   
60        2020  CP063   
61        2020  CP064   
62        2020  CP065   
63        2020  CP066   
64        2020  CP067   
65        2020  CP068   
66        2020  CP069   
67        2020  CP070   
68        2020  CP071   
69        2020  CP072   
70        2020  CP073   
71        2020  CP074   
72        2020  CP075   

                                                                                                           Project_Title  \
0                                                         Regional Transit Interconnectivity & Environmental Sustability   
1                                                                                          Travel Time Reduction Project   
2                                 Willowbrook/Rosa Parks Station & Blue Line Light Rail Operational Improvements Project   
3                                                                             Pacific Surfliner Transit Transfer Program   
4                                       Monterey Bay Operations and Maintenance Facility/Salinas Transit Service Project   
5                                                                                           Bravo! Route 560 Rapid Buses   
6                                                   Sacramento Regional Transit's Refurbishment of 7 Light Rail Vehicles   
7                                                                                            South Bay Bus Rapid Transit   
8                                                    San Diego Metropolitan Transit System Trolley Capacity Improvements   
9                                                                               SFMTA Light Rail Vehicle Fleet Expansion   
10                                                                                             ACE Wayside Power Project   
11                                          Bus Rapid Transit – Martin Luther King Corridor and Crosstown Miner Corridor   
12                                                         Purchase of Nine Fuel-Efficient Tier IV EMD F-125 Locomotives   
13                                                                                       SMART Rail Car Capacity Project   
14                  Zero Emission Bus and Vanpool Expansion in the Antelope Valley, Kern County and the Coachella Valley   
15                               Increased Rail Service to Roseville, Service Optimization and Standby Power Investments   
16                                  Transform

## Tableau

In [40]:
def tableau2(df):
    #Keeping only the columns we want
    df = (df[['PPNO','Award_Year', 'Project_#', 'Grant_Recipient', 
       'Project_Title', 'District', 'County', 'Project_Description',
       'Master_Agreement_Number', 'Master_Agreement_Expiration_Date',
       'Project_Manager', 'Regional_Coordinator', 'Technical_Assistance-Fleet_(Y/N)',
       'Technical_Assistance-Network_Integration_(Y/N)',
       'Technical_Assistance-Priority_Population_(Y/N)', 'Total_Project_Cost',
       'TIRCP_project_sheet', 'Allocated_Amount',
       'Percentage_Allocated',
       'from_joined_alloc_df_tot_exp', 'Other_Funds_Involved']]
                 )
    #Getting percentages & filling in with 0
    df['Expended_Percent'] = df['from_joined_alloc_df_tot_exp']/df['Allocated_Amount']
    df['Allocated_Percent'] = df['Allocated_Amount']/df['TIRCP_project_sheet']
    
    # Subtract TIRCP with Allocated Amount with Unallocated
    df['Unallocated_Amount'] = df["TIRCP_project_sheet"] - df["Allocated_Amount"]
    #filling in for 0's 
    df[['Expended_Percent','Allocated_Percent']] = df[['Expended_Percent','Allocated_Percent']].fillna(value=0)
   
    df[['Expended_Percent','Allocated_Percent']]  =  df[['Expended_Percent','Allocated_Percent']].replace(np.inf, 0)
    #Categorizing expended percentage into bins
    def expended_percent(row):
           
            if ((row.Expended_Percent > 0) and (row.Expended_Percent < .26)):
                return "1-25"
            elif ((row.Expended_Percent > .25) and (row.Expended_Percent < .51)):
                return "26-50"
            elif ((row.Expended_Percent > .50) and (row.Expended_Percent < .76)):
                return "51-75"
            elif ((row.Expended_Percent > .75) and (row.Expended_Percent < 1.0)):
                return "76-99"
            elif row.Expended_Percent == 0.0:
                return "0"
            else:
                return "100"
    df["Expended_Percent_Group"] = df.apply(lambda x: expended_percent(x), axis=1)
    
    # Categorize years and expended_percent_group into bins
    def progress(df):   
        ### 2015 ### 
        if (df['Award_Year'] == 2015) and (df['Expended_Percent_Group'] == "1-25") | (df['Expended_Percent_Group'] == "26-50"):
            return 'Behind'
        elif (df['Award_Year'] == 2015) and (df['Expended_Percent_Group'] == "76-99") | (df['Expended_Percent_Group'] == "51-75"):
            return 'On Track'
        
        ### 2016 ###
        elif (df['Award_Year'] == 2016) and (df['Expended_Percent_Group'] == "1-25") | (df['Expended_Percent_Group'] == "26-50"):
            return 'Behind'
        elif (df['Award_Year'] == 2016) and (df['Expended_Percent_Group'] == "51-75") | (df['Expended_Percent_Group'] == "76-99"):
            return 'On Track'
        
        ### 2018 ###
        elif (df['Award_Year'] == 2018) and (df['Expended_Percent_Group'] == "1-25"):
            return 'Behind'
        elif (df['Award_Year'] == 2018) and (df['Expended_Percent_Group'] == "26-50") | (df['Expended_Percent_Group'] == "51-75"):
            return 'On Track'
        elif (df['Award_Year'] == 2018) and (df['Expended_Percent_Group'] == "76-99"):
            return 'Ahead'
        
        ### 2020 ### 
        elif (df['Award_Year'] == 2020) and (df['Expended_Percent_Group'] == "1-25"):
            return 'Behind'
        elif (df['Award_Year'] == 2020) and (df['Expended_Percent_Group'] == "26-50"):
            return 'On Track'
        elif (df['Award_Year'] == 2020) and (df['Expended_Percent_Group'] == "51-75") | (df['Expended_Percent_Group'] == "76-99"):
            return 'Ahead'
        
        ### 0 Expenditures ### 
        elif df ['Expended_Percent_Group'] == "0":
            return "No expenditures recorded"
        
        ### Else ### 
        else: 
            return "100% of allocated funds spent"

    df['Progress'] = df.apply(progress, axis = 1)
    
    #Which projects are large,small, medium
    p75 = df.TIRCP_project_sheet.quantile(0.75).astype(float)
    p25 = df.TIRCP_project_sheet.quantile(0.25).astype(float)
    p50 = df.TIRCP_project_sheet.quantile(0.50).astype(float)
    
    def project_size (row):
        if ((row.TIRCP_project_sheet > 0) and (row.TIRCP_project_sheet < p25)):
            return "Small"
        elif ((row.TIRCP_project_sheet > p25) and (row.TIRCP_project_sheet < p75)):
            return "Medium"
        elif ((row.TIRCP_project_sheet > p50) and (row.TIRCP_project_sheet > p75 )):
            return "Large"
        else:
            return "$0 recorded for TIRCP"
        
    df["Project_Category"] = df.apply(lambda x: project_size(x), axis=1)
     #Rename cols to the right names
    df = (df.rename(columns = {'from_joined_alloc_df_tot_exp':'Expended_Amount', 
                                                'TIRCP_project_sheet': "TIRCP_Amount"}
                  ))
    ### GCS ###
    # df = df.to_excel(f'{GCS_FILE_PATH}INTERIM_EXPENDITURE_Tableau_Parquet.xlsx')
  
    return df 

In [41]:
final_df = tableau2(final_join)

/tmp/ipykernel_1219/964883781.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/tmp/ipykernel_1219/964883781.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/tmp/ipykernel_1219/964883781.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/tmp/ipykernel_1219/96488

In [42]:
final_df[['Award_Year','Project_#','PPNO','Project_Title', 'Allocated_Amount','TIRCP_Amount',
          'Expended_Amount','Expended_Percent','Progress']].sort_values( by = ['Award_Year', 'PPNO'])

Award_Year  Project_#   PPNO  \
6         2015          7  CP001   
12        2015         13  CP002   
7         2015          8  CP003   
5         2015          6  CP004   
0         2015          1  CP005   
9         2015         10  CP006   
3         2015          4  CP007   
8         2015          9  CP008   
13        2015         14  CP010   
11        2015         12  CP011   
1         2015          2  CP012   
4         2015          5  CP013   
10        2015         11  CP014   
2         2015          3  CP015   
24        2016         11   1230   
25        2016         12  CP006   
17        2016          4  CP016   
21        2016          8  CP017   
14        2016          1  CP019   
19        2016          6  CP020   
22        2016          9  CP022   
15        2016          2  CP023   
26        2016         13  CP025   
20        2016          7  CP043   
27        2016         14  CP057   
18        2016          5  CP058   
16        2016          3  CP076   
23        2016         10  CP080   
55        2018         28   1155   
28        2018          1  2320B   
45        2018         18  CP006   
29        2018          2  CP027   
30        2018          3  CP028   
34        2018          7  CP029   
37        2018         10  CP030   
38        2018         11  CP031   
39        2018         12  CP031   
54        2018         27  CP033   
42        2018         15  CP034   
44        2018         17  CP034   
46        2018         19  CP035   
32        2018          5  CP036   
40        2018         13  CP039   
53        2018         26  CP041   
51        2018         24  CP045   
52        2018         25  CP046   
49        2018         22  CP047   
48        2018         21  CP048   
36        2018          9  CP051   
41        2018         14  CP053   
35        2018          8  CP054   
31        2018          4  CP055   
50        2018         23  CP057   
43        2018         16  CP077   
47        2018         20  CP078   
33        2018          6  CP079   
56        2020          1  CP059   
57        2020          2  CP060   
58        2020          3  CP061   
59        2020          4  CP062   
60        2020          5  CP063   
61        2020          6  CP064   
62        2020          7  CP065   
63        2020          8  CP066   
64        2020          9  CP067   
65        2020         10  CP068   
66        2020         11  CP069   
67        2020         12  CP070   
68        2020         13  CP071   
69        2020         14  CP072   
70        2020         15  CP073   
71        2020         16  CP074   
72        2020         17  CP075   

                                                                                                           Project_Title  \
6                                                   Sacramento Regional Transit's Refurbishment of 7 Light Rail Vehicles   
12                                                         Purchase of Nine Fuel-Efficient Tier IV EMD F-125 Locomotives   
7                                                                                            South Bay Bus Rapid Transit   
5                                                                                           Bravo! Route 560 Rapid Buses   
0                                                         Regional Transit Interconnectivity & Environmental Sustability   
9                                                                               SFMTA Light Rail Vehicle Fleet Expansion   
3                                                                             Pacific Surfliner Transit Transfer Program   
8                                                    San Diego Metropolitan Transit System Trolley Capacity Improvements   
13                                                                                       SMART Rail Car Capacity Project   
11                                          Bus Rapid Transit – Martin Luther King Corridor and Cro

In [43]:
final_df.Progress.value_counts()

No expenditures recorded         24
On Track                         18
Behind                           14
100% of allocated funds spent    12
Ahead                             5
Name: Progress, dtype: int64

In [44]:
final_df.Progress.value_counts().sum()

73

In [45]:
TIRCP = final_df['TIRCP_Amount'].sum()
f'The total amount of TIRCP is {TIRCP}'

'The total amount of TIRCP is 5440171000'

In [46]:
Alloc = final_df['Allocated_Amount'].sum()
f'The total amount of $ allocated is {Alloc}'

'The total amount of $ allocated is 2266402000'

In [47]:
Exp = final_df['Expended_Amount'].sum()
f'Total amount of expenditures is {Exp}'

'Total amount of expenditures is 743099838.47'

In [48]:
f'% of allocated money spent {(Alloc/TIRCP).round(2)}'

'% of allocated money spent 0.42'

In [49]:
f'% of expended  by allocation {(Exp/Alloc).round(2)}'

'% of expended  by allocation 0.33'

In [50]:
f'% of expended  by TIRCP {(Exp/TIRCP).round(2)}'

'% of expended  by TIRCP 0.14'

In [51]:
final_df.to_excel(f'{GCS_FILE_PATH}tableau_with_temporary_expenditure_sol.xlsx', index = False)